File: src/main/scala/category/twopointer/_424_Solution.scala
package category.twopointer

/**
  * 424. Longest Repeating Character Replacement
  */
object _424_Solution {
  def characterReplacement(s: String, k: Int): Int = {
    var p0, p1, maxCnt, maxLength = 0
    val bucket = new Array[Int](26)
    while (p1 < s.length) {
      maxCnt = Math.max(maxCnt, {bucket(s(p1) - 'A') += 1; bucket(s(p1) - 'A')})
      if (p1 - p0 + 1 - maxCnt > k) {bucket(s(p0) - 'A') -= 1; p0 += 1}
      maxLength = Math.max(maxLength, p1 - p0 + 1)
      p1 += 1
    }
    maxLength
  }
  
  //  def characterReplacement(s: String, k: Int): Int = {
  //    var p0, p1, max = 0
  //    while (p0 < s.length) {
  //      var remain = k
  //      val base = s(p0)
  //      val start = p0
  //      while (p0 < s.length && s(p0) == base) p0 += 1
  //      p1 = p0
  //      while (remain >= 0 && p1 < s.length) {
  //        if (s(p1) != base) remain -= 1
  //        p1 += 1
  //      }
  //      max = Math.max(max, p1 + remain - start)
  //    }
  //    Math.min(s.length, max)
  //  }
}

File: src/main/python/medium/_400_450/_424_Solution.py
# 424. Longest Repeating Character Replacement
'''
"ABAB"
2
"AABABBA"
2
"AABABBA"
1
""
1
"AA"
3
"CABDEFGHTYQQW"
1
"AAAAAAAAAA"
1
"ABABABABABAAB"
1
"ABCCC"
1
"A"
0
"ABCC"
0
"BAAAB"
2
'''


class _424_Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        word = [0] * 26
        i, j = 0, 0
        maximal, max_char = 0, 0
        while j < len(s):
            c = s[j]
            word[ord(c) - ord('A')] += 1
            all = j - i + 1
            max_char = max(max_char, word[ord(c) - ord('A')])
            remain = all - max_char
            if remain <= k:
                maximal = max(maximal, all)
                j += 1
            else:
                word[ord(s[i]) - ord('A')] -= 1
                i += 1
                j += 1
        return maximal


if __name__ == '__main__':
    obj = _424_Solution()
    assert obj.characterReplacement("BAAAB", 2) == 5
    assert obj.characterReplacement("ABABABABABAAB", 1) == 4
    assert obj.characterReplacement("ABAB", 2) == 4
    assert obj.characterReplacement("AABABBA", 1) == 4
